File: algorithms/29.cpp
#include <iostream>
#include <climits>
#include <cmath>
using namespace std;

class Solution
{
public:
    int divide(int dividend, int divisor)
    {
        if( divisor == 0 || (dividend == INT_MIN && divisor == -1) )
            return INT_MAX;
        bool sign = (dividend<0)==(divisor<0);
        unsigned int a = abs(dividend), b = abs(divisor);
        int ans = 0;
        for ( int i = 31; i >= 0; i-- )
        {
            if ( a>>i >= b )
            {
                ans = (ans<<1)|1;
                a -= (b<<i);
            }
            else
                ans = ans<<1;
        }
        if ( sign )
            return ans;
        else
            return -ans;
    }
};

int main()
{
    Solution solution;
    cout << solution.divide(5, 2) << endl;
    return 0;
}

File: algorithms/rust/29.rs
impl Solution {
    #[allow(unused)]
    pub fn divide(dividend: i32, divisor: i32) -> i32 {
        if dividend == i32::MIN && divisor == -1 {
            return i32::MAX;
        }
        let neg = (dividend ^ divisor) < 0;
        let mut dividend = dividend.unsigned_abs();
        let divisor = divisor.unsigned_abs();

        let mut quotient = 0;
        for i in (0..32).rev() {
            if ((dividend >> i) >= divisor) {
                quotient += 1 << i;
                dividend -= divisor << i;
            }
        }
        return if neg { -quotient } else { quotient };
    }
}

#[allow(dead_code)]
struct Solution;

fn main() {
    println!("Hello, world!");
}
